Add DetectorData API

philsmt opened the pull request:
As the CAL team has been refactoring the injection code away from the legacy `calibrationDBRemote` API, we noticed that handling PDU details is somewhat cumbersome and requires replicating non-obvious RESTful requests in many places. As we do not want to add further auxiliary functions to `CalCatAPIClient`, this PR adds an API accompanying `CalibrationData` by `DetectorData`.

philsmt committed:
Add DetectorData API

takluyver reviewed (commented):
Is the goal to replace `CalibrationData.module_details` and `.detector_name` with a `DetectorData` object? If so, we might need to do something about the `.from_correction(..., use_calcat=False)` case, where we don't have most of the info this assumes, just the bare minimum of detector name, aggregator names and PDU names.

takluyver commented on src/extra/calibration.py:
```diff
@@ -1329,6 +1293,199 @@ class ShimadzuHPVX2Conditions(ConditionsBase):
+
+    PDUs describe the physical modules independent from the detector
+    they may be installed in, or the detector's logical modules they are
+    mapped to.
```
How about we call this class something like `DetectorModule`, and rework the docstring to say that it describes a module installed in a detector at a particular point in time? It's misleading to say it's "independent from the detector [it] may be installed in" when it has fields describing the detector name and the module's position in it.

I guess a more idealised approach would be to split the class up into `PhysicalDetectorUnit` (with id, name, uuid) and `InstalledModule` with a reference to a PDU plus the mapping fields. But I don't think we really need that.

takluyver commented on src/extra/calibration.py:
```diff
@@ -1329,6 +1293,199 @@ class ShimadzuHPVX2Conditions(ConditionsBase):
+    detector: str  # Detector identifier the PDU is currently mapped to
+    virtual_device_name: str  # Identifier within the detector, e.g. Q1M2
+    module_index: int  # Enumerated module index within the detector, contiguous and always starts at 0
+    module_number: int  # Module number within the detector, may start at any number and have gaps
```
Make it explicit, which can also be written `int | None` since Python 3.10. Labelling a field as just `int` implies that it is always an int.

takluyver commented on src/extra/calibration.py:
```diff
@@ -1329,6 +1293,199 @@ class ShimadzuHPVX2Conditions(ConditionsBase):
+    @property
+    def identifier(self) -> str:
+        """Detector identifier."""
+        return self.detector_row['identifier']
```
I'd mostly make them regular attributes, i.e. `self.identifier = d['identifier']`. I think wrapping attributes into read-only properties makes the code longer for little real purpose.

 If it's useful to to e.g. throw an error rather than exposing `None`, like you've done for `source_name_pattern`, then yes, I think a private attribute wrapped in a property is the way to do it. But 'None if not set' is a very common pattern, and I'm not sure that an AssertionError is really clearer. 🤷

philsmt committed:
Implement CalibrationData on top of DetectorData

philsmt commented:
Apart from addressing the comments, I tried to base the `CalibrationData` on top of `DetectorData`.

I tried to do this in a minimalistic approach by preserving all business logic and *only* emulating `detector_name` and `module_details` underneath. It passes the test suite just fine, what are your thoughts?

Technically there are some differences in obscure fields like `detector_type_id` and such, but I don't expect this to matter in practice.

takluyver commented on tests/test_calibration.py:
```diff
@@ -376,3 +378,76 @@ def test_conditions_markdown():
+
+@pytest.mark.vcr
+def test_DetectorData_list_by_instrument():
+    assert DetectorData.list_by_instrument('SCS') == [
```
I don't think this can be used with a snapshot time at the moment, but maybe we can ask if that's easy to add.

philsmt replied:
Yes, I double-checked the detector APIs and indeed they do not have snapshotting in those APIs at the moment.

takluyver commented on src/extra/calibration.py:
```diff
@@ -1329,6 +1367,257 @@ class ShimadzuHPVX2Conditions(ConditionsBase):
+
+            self.pdus.append(item)
+
+        self.pdu_snapshot_at = pdu_snapshot_at or datetime.now().isoformat()
```
IIRC this was in the actual constructor, which is being passed PDU info. It makes sense to default to *retrieving* current state if no timestamp is specified, but not to assume the information you've got from somewhere else is the current state. Looking at the current code, I think you've fixed this now, though.

philsmt committed:
Remove DetectorData.source_names for now

philsmt commented on src/extra/calibration.py:
```diff
@@ -1329,6 +1367,257 @@ class ShimadzuHPVX2Conditions(ConditionsBase):
+    def source_names(self) -> list[str]:
+        """Source names of currently mapped PDUs."""
+        return [self.source_name_pattern.format(
+            modno=pdu.module_number or i + self.first_module_index
```
Good point, I removed it via an explicit commit, so we can easily get back later.

philsmt commented:
I cleaned up the history now, could you please have a look at the remaining discussions on whether they're actionable?

takluyver commented on tests/test_calibration.py:
```diff
@@ -376,3 +378,80 @@ def test_conditions_markdown():
+        'AGIPD_SIV1_AGIPDV11_M517', 101003000000, 'AGIPD-Type')
+
+    # FXE-JFHZ, single-module detector with partial CalCat entries
+    jfhz = DetectorData.from_identifier('FXE_XAD_JFHZ')
```
Should this have the timestamp?

```suggestion
    jfhz = DetectorData.from_identifier('FXE_XAD_JFHZ', **pdu_date_kw)
```

takluyver committed:
Re-record VCR cassettes for failing tests
